File: leetcode/python/000297.serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left: TreeNode | None = None
        self.right: TreeNode | None = None


class Codec:

    # Time complexity: O(n)
    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        s = []

        def dfs(node):
            if node is None:
                s.append("N")
                return

            s.append(str(node.val))
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return ",".join(s)

    # Time complexity: O(n)
    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        values = data.split(",")
        self.index = 0

        def dfs():
            if values[self.index] == "N":
                self.index += 1
                return None

            node = TreeNode(int(values[self.index]))
            self.index += 1
            node.left = dfs()
            node.right = dfs()

            return node

        return dfs()


# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: leetcode/typescript/000297.serialize-and-deserialize-binary-tree.ts
/**
 * Definition for a binary tree node.
 */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/*
 * Encodes a tree to a single string.
 */
function serialize(root: TreeNode | null): string {
  const res: Array<"N" | number> = [];

  function dfs(node: TreeNode | null): void {
    if (!node) {
      res.push("N");
      return;
    }
    res.push(node.val);
    dfs(node.left);
    dfs(node.right);
  }

  dfs(root);
  return res.join(",");
}

/*
 * Decodes your encoded data to tree.
 */
function deserialize(data: string): TreeNode | null {
  const values = data.split(",");
  let i: number = 0;

  function dfs() {
    if (values[i] === "N") {
      i++;
      return null;
    }
    const node = new TreeNode(Number(values[i]));
    i++;
    node.left = dfs();
    node.right = dfs();
    return node;
  }

  return dfs();
}

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */
